### PR TITLE
Adding `is_ai_graded` column to `grading_jobs`

### PR DESCRIPTION
**Pull request by Victorsss-Orz**
#10717 

Add a column to `grading_jobs` to indicate if a submission is ai graded. This would prevent RAG from training AI using AI generated data.

### CODE REVIEW

**Comment by Victorsss-Orz** [comment]
@jonatanschroeder I made the changes to only `grading_job` for now, and modified the `updateInstanceQuestionScore` accordingly to add a default for `is_ai_graded`. If you would like to take a quick look that would be great.

**Comment by Victorsss-Orz** [comment]
@nwalters512 Is there a way to run the query planner on the real PL database for me? I've modified the `select_closest_embeddings` block to account for the `is_ai_graded` information, and the performance is lowered by about 30% compared to the original version. Can I assume that in production, the performance change will be about the same (e.g. from 2ms to about 2.6ms)?

And I would also need to show an additional column in the instructor assessment question manual grading page. I'm wondering maybe I should leave that to a later PR so the PRs are broken down to smaller parts and make the reviews easier.

**Comment by nwalters512** [comment]
@Victorsss-Orz I checked out these queries for you against the production database. The most expensive part is still loading all the relevant rows into cache. I checked data against a course with ~450 students. For a given assessment question and a cold cache, this PR roughly doubles the time this query takes (~2.5s to ~5s). Once data is in cache, performance is much better (~15ms).

While I wish the initial query were more performant and didn't drag _quite_ so much data into memory, I think this is acceptable for research code, and the fact that the query is quite fast once things _are_ in memory makes me feel better about this change.

File: apps/prairielearn/src/ee/lib/ai-grading.sql
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/ai-grading.sql:109]: There are two ways we could implement filtering here:

- For each instance question, use the latest non-AI-graded grading job
- For each instance question, use the latest grading job only if it is non-AI-graded

These are subtly different. You implemented the second option, is that the one we want here? Not asking for a change, only a confirmation that you've thought this through.

File: apps/prairielearn/src/ee/lib/ai-grading.ts
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/ai-grading.ts:378]: Like line 371, it'd be nice to have a trailing comment stating what this is:

```suggestion
            true, // is_ai_graded
```

This function is up to 7 arguments, so it may be a good candidate to switch from individual arguments to an object with named properties. I won't request that change here. cc @jonatanschroeder, this might be a nice refactor sooner or later.

**Review by nwalters512 — APPROVED** [review]
LGTM. Please confirm that you're filtering grading jobs in the way that makes most sense to you.

File: apps/prairielearn/src/ee/lib/ai-grading.sql
Comment (Victorsss-Orz) [reply apps/prairielearn/src/ee/lib/ai-grading.sql:109]: Yes. The second is what I've intended to build here.

Comment (Victorsss-Orz) [reply apps/prairielearn/src/ee/lib/ai-grading.sql:109]: My rationale, just in case people are wondering: 
For manual grades, it is hard for instructors to check the older manual grading results for the same submission once it is regraded (from what I've tried with manual grading), so the quality of non-latest grading jobs are not guaranteed. If I give a random and bad grade to a submission first, and then overwrite it with an AI grade, the older bad grade would be used if I'm using the latest non-AI-graded job.

**Comment by Victorsss-Orz** [comment]
> @Victorsss-Orz I checked out these queries for you against the production database. The most expensive part is still loading all the relevant rows into cache. I checked data against a course with ~450 students. For a given assessment question and a cold cache, this PR roughly doubles the time this query takes (~2.5s to ~5s). Once data is in cache, performance is much better (~15ms).
> 
> While I wish the initial query were more performant and didn't drag _quite_ so much data into memory, I think this is acceptable for research code, and the fact that the query is quite fast once things _are_ in memory makes me feel better about this change.

I'm really struggling to find better ways to do this, including for the earlier version... I would definitely need to find one submission for each instance question, which aren't happening without going through `instance_questions`, `variants`, and then `submissions`. Or maybe there can be shortcuts that I'm not aware of?

**Comment by nwalters512** [comment]
@Victorsss-Orz I really don't think there is a better option at this point. If this ends up being a major bottleneck (I'm not sure if it will, and it certainly won't til this is exposed to more users), we maybe be able to improve things with denormalizing data (adding more `id` columns to things closer to the leaves of our relation graph), adding more indexes, or both.

My approval stands. I'm going to see if Matt wants to take a look at this. If he does, I'll wait for him to do so; if not, I'll go ahead and merge.

**Comment by mwest1066** [comment]
My current thinking is that we should also add `is_ai_graded` to `submissions` and `instance_questions`. Scores trickle up from `grading_jobs` to `submissions` and hence to `instance_questions`, and we would also flow up `is_ai_graded` each time we use a grading job. This would avoid needing to query grading jobs directly and I feel like it matches our mental model so that the answer to "was this question graded by an AI?" is always "was the latest grade assigned to it assigned by an AI?".

**Comment by Victorsss-Orz** [comment]
I agree that adding the column along the hierarchy of tables would be a good way of preventing the need to join so many tables, if this is implemented. However, I currently have almost 0 idea on how to achieve this.

**Comment by jonatanschroeder** [comment]
https://github.com/PrairieLearn/PrairieLearn/issues/10717#issuecomment-2386084502

**Comment by Victorsss-Orz** [comment]
I think the problem on hand is performance of the queries. By adding the info into `instance_question` and `submission`, we don't have to join so many tables just to get this `is_ai_graded` column.

**Comment by jonatanschroeder** [comment]
Do you need instance_questions? Isn't submissions enough?

**Comment by Victorsss-Orz** [comment]
Later I would need to show the `is_ai_graded` info on the instance question level so instructor can see if the instance question is graded by a human or AI in the manual grading assessment question page. The only options for this is to either have the information in the `instance_questions` table, or joining both `variant` and `submissions` (and potentially `grading_jobs`) onto the table to retrieve that one column, which would be really low performing.

**Comment by mwest1066** [comment]
> Do you need instance_questions? Isn't submissions enough?

My thinking here was that it useful to know whether the last "manual" grading update to a question was done by a human or by AI. While it's true that manual grading is always associated with particular submissions, and we could join over the latest submission, it seems to me to be simpler to denormalize this and bubble up the information to instance questions. We are already bubbling up the score information so there isn't any more work here, and it will simplify later queries that want to show which questions were human/AI graded.

@jonatanschroeder is there any particular disadvantage that you can see with storing this in instance questions, or are you mainly just thinking that it isn't necessary?

**Comment by jonatanschroeder** [comment]
I'm ok with that approach.

**Comment by Victorsss-Orz** [comment]
I'll close this PR for now, and will open another one when I start to actually work on this.

**Pull request closed without merging**
